### ISSUE TITLE
Compile step from README fails creating initial release

### ISSUE THREAD

**Issue by MaxPower15** [boldpoker/edeliver#73]
I was following the README instructions to test out a deployment using edeliver all on my local machine. I'm using elixir 1.2.0, edeliver 1.2.4, and exrm 1.0.4, running on OSX. This is the output:

```
mix edeliver build release

BUILDING RELEASE OF WEBHOOKS APP ON BUILD HOST

-----> Authorizing hosts
-----> Ensuring hosts are ready to accept git pushes
-----> Pushing new commits with git to: maxschnur@127.0.0.1
-----> Resetting remote hosts to 3a545059f53b2fc07436acffd97c8758772bc93a
-----> Cleaning generated files from last build
-----> Fetching / Updating dependencies
-----> Compiling sources

== Compilation error on file lib/webhooks/event.ex ==
** (ArgumentError) Poison.Encoder is not available, cannot derive Poison.Encoder for Webhooks.Event
    (elixir) lib/protocol.ex:69: Protocol.assert_protocol!/2
    (elixir) lib/protocol.ex:582: Protocol.derive/5
    (stdlib) lists.erl:1337: :lists.foreach/2
    (elixir) lib/protocol.ex:570: Protocol.__derive__/3
    lib/webhooks/event.ex:40: (module)


== Compilation error on file lib/event_poster/handler.ex ==
** (CompileError) lib/event_poster/handler.ex:7: Webhooks.Event.__struct__/0 is undefined, cannot expand struct Webhooks.Event
    (elixir) src/elixir_map.erl:58: :elixir_map.translate_struct/4
    (stdlib) lists.erl:1353: :lists.mapfoldl/3


== Compilation error on file lib/event_poster.ex ==
** (CompileError) lib/event_poster.ex:20: NSQ.Config.__struct__/0 is undefined, cannot expand struct NSQ.Config
    (elixir) src/elixir_map.erl:58: :elixir_map.translate_struct/4
    lib/event_poster.ex:19: (module)

FAILED 1:

    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd .deliver
    if [ "mix" = "rebar" ]; then
      echo "using rebar to compile files"
      [[ "" != "true" ]] && ./rebar  clean skip_deps=true || :
      ./rebar  compile
    elif [ "mix" = "mix" ] && [ "mix" = "mix" ]; then
      if [[ "" = "true" ]]; then
        APP="webhooks" MIX_ENV="prod" AUTO_VERSION="" mix do compile
      else
        APP="webhooks" MIX_ENV="prod" AUTO_VERSION="" mix do clean, compile
      fi
    elif [ "mix" = "mix" ]; then
      echo "using mix to compile files"
      if [[ "" = "true" ]]; then
        APP="webhooks" MIX_ENV="prod" AUTO_VERSION="" mix do deps.compile, compile
      else
        APP="webhooks" MIX_ENV="prod" AUTO_VERSION="" mix do clean, deps.compile, compile
      fi
    fi
```

I did some testing locally and found that the command `APP="webhooks" MIX_ENV="prod" AUTO_VERSION="" mix do clean, compile` was the only one throwing that error.

And this is my config file:

```
#!/usr/bin/env bash
APP="webhooks" # name of your release
BUILD_HOST="127.0.0.1" # host where to build the release
BUILD_USER="maxschnur" # local user at build host
BUILD_AT=".deliver" # build directory on build host
RELEASE_DIR="/opt/webhooks" # deploy directory on production hosts
```

If I run `SKIP_MIX_CLEAN=true mix edeliver build release`, compilation succeeds but it then fails at the "Generating release" step with this error:

```
zsh:2: no matches found: /opt/webhooks/releases/*/
Failed to detect generated release version at:
maxschnur@127.0.0.1:/opt/webhooks/releases/

No built release found.
```

Something seems off here... if this is the first release, why does it need to find an existing built release? From what I can tell, this error happens because there are no folders in /opt/webhooks. Manually creating an empty folder with any name in that directory fixes the error. Alternatively, explicitly setting the RELEASE_VERSION environment variable also fixes it.

I should mention that this error was after getting an error about RELEASE_DIR not being set. I added that to my config to get around the issue, but I didn't see that in the example config or in the example command-line, where it would be valid as an environment variable.

So, overall I was able to work around this, but I'm unsure of what the correct behavior is actually supposed to be. Should I be manually defining RELEASE_VERSION on first build? Should I need to set SKIP_MIX_CLEAN? Is there some other configuration I've messed up? Is what I did expected and it's just documentation that needs an update?

**Comment by bharendt** [comment]
Based on your detailed description of the error, of your setup and your approach, it looks like the follwing things happened:

 - The compilation errors indicate that some deps are misssing, e.g poison
 - That's why the errors occur in the `mix do clean, compile` step when compiling the sources for the release
 - If compiliation succeed when you were using `SKIP_MIX_CLEAN=true`, this actually means that you somehow fixed the compilation issue manually before, because this flag just causes edeliver to compile only changed sources instead of recompiling all sources. edeliver then used your existing beam files (from local testing / fixing) to generate the release.
 
So the main question to solve the actual issue would be: what did you do (when testing locally) to solve the compilation error?

But I try to answer also your other questions: Usually you should not need to set the `RELEASE_DIR` config, but maybe the README and/or the error message is misleading in that case. Usually the `RELEASE_DIR` is autodetected by edeliver and represents the directory where the generated release is located. This depends on the build tool (exrm/rebar etc.), but should be something like `$BUILD_AT/rel/$APP` in your case (containging the `$APP.tar.gz`) and has nothing to do with the `$DELIVER_TO` directory, which represents the directory where you deploy the built release on the production host(s) (in your case maybe `/opt/webhooks`).
So if the built release (`webhooks.tar.gz`) cannot be found at `.deliver/rel/webhooks` usually it either was not built or there are serveral releases (from a previous build) and edeliver does not know which is the current one. If setting `RELEASE_VERSION` helped in your case, this indicates that the latter one was the case. This can be caused by using the `--skip-git-clean` option (or `SKIP_GIT_CLEAN=true`) or wrongly adjusting the `GIT_CLEAN_PATHS=` option, e.g. by forgetting to add the directory containing the built releases. 

In summary:

 - it would good to know how you solved the compilation issue
 - you should not need to set `RELEASE_DIR` in the config, and if setting it for one build it represents the directory where the built release can be found, not where it should be deployed on the deploy host(s)
 - you should not need to set `RELEASE_VERSION` in the config. 
 - if several built releases are found and setting `RELEASE_VERSION` solves to find the correct one, maybe `--skip-git-clean` (or `SKIP_GIT_CLEAN`) was used and should not be used any more or `GIT_CLEAN_PATHS=` should be adjusted to contain the `RELEASE_DIR`

Hope this helps and let us know if you have some new information.

**Comment by bharendt** [comment]
@MaxPower15: Do you have any new information that might help to solve this issue or that might help others with similar problems? Or should I close this issue?

**Comment by MaxPower15** [comment]
Hi @bharendt -- sorry, I've been focused on other things recently and won't have time to put more research time into this for at least a couple more weeks. From my perspective, this is still an open issue. But if you'd like to keep this area clean, you can close and I'll reopen when I have more info.

**Comment by bharendt** [comment]
@MaxPower15: No problem, then I will keep it open. Just let us know if you have some new information and we will try to help you to solve this problem.

**Comment by MaxPower15** [comment]
Way late on the follow-up for this one, sorry! I hacked around with this today and got it working. I think the root issue for me was a combination of misunderstanding what happens at each step of build/deploy, the config for that (thanks for the hints), and a disconnect between my mix.exs and .deliver/config files. That is, they had different application names, the mix.lock file was mistakenly placed outside the git repo. The combination of these things confused me for quite a while, heh. Should've tried this with a clean repo to start. :P

Anyway, thanks for your patience on this one. Closing!
